### PR TITLE
Fix code generators-- make scheme building composable

### PR DESCRIPTION
**Pull request by lavalamp**
I needed to make some changes to make my other refactoring possible and this got rather large. 

We now provide a "SchemeBuilder" to help all of the api packages provide their scheme-building functions (addKnownTypes and friends) in a standardized way. This also allows generated deepcopies & conversions to be entirely self contained, the project will now build without them being present (as they can add themselves to the SchemeBuilder). (Although if you actually build without them, you will get reduced performance!)

Previously, there was no way to construct your own runtime.Scheme (e.g., to test), you had to use the api.Scheme object, which has all sorts of non-hermetic cruft in it. Now you can get everything from a package by calling the scheme builder's AddToScheme, including the generated functions, if they are present.

Next steps are to allow for declaring dependencies, and to standardize the registration & install code. (#25434)

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/25526)

### CODE REVIEW

File: cmd/libs/go2idl/conversion-gen/generators/conversion.go [hunk 1]
```diff
@@ -499,12 +499,13 @@ func (g *genConversion) preexists(inType, outType *types.Type) (*types.Type, boo
 
 func (g *genConversion) Init(c *generator.Context, w io.Writer) error {
 	sw := generator.NewSnippetWriter(w, c, "$", "$")
-	sw.Do("func init() {\n", nil)
-	if g.targetPackage == apiPackagePath {
-		sw.Do("if err := Scheme.AddGeneratedConversionFuncs(\n", nil)
-	} else {
-		sw.Do("if err := api.Scheme.AddGeneratedConversionFuncs(\n", nil)
-	}
+	sw.Do("func AddGeneratedConversionFuncs(scheme *$.|public$) {\n",
```
Comment (wojtek-t) [inline cmd/libs/go2idl/conversion-gen/generators/conversion.go]: I'm pretty sure Clayton was doing similar changes - maybe this PR is not yet merged.

Comment (lavalamp) [inline cmd/libs/go2idl/conversion-gen/generators/conversion.go]: @smarterclayton - am I duplicating effort? :/

**Comment by lavalamp** [comment]
Prereq for #25434.

Passes unit tests locally now.

**Comment by lavalamp** [comment]
Ready for review.

**Comment by lavalamp** [comment]
hm, maybe not.

File: plugin/pkg/scheduler/api/v1/register.go [hunk 2]
```diff
@@ -26,7 +27,16 @@ import (
 var SchemeGroupVersion = unversioned.GroupVersion{Group: "", Version: "v1"}
 
 func init() {
-	api.Scheme.AddKnownTypes(SchemeGroupVersion,
+	addKnownTypes(api.Scheme)
```
Comment (smarterclayton) [inline plugin/pkg/scheduler/api/v1/register.go]: Ironically this makes it harder to understand what is happening here (we add types on init, create a scheme builder that calls a local method as well, then define a public var that doesn't point to the local method except via indirection.  I understand it, it's just full of twisty packages.

Comment (lavalamp) [inline plugin/pkg/scheduler/api/v1/register.go]: Yes, I am absolutely not done fixing this code, it's super inconsistent right now. I want to remove all of these init functions, but I left them in so as not to change behavior yet.

**Comment by lavalamp** [comment]
Fixed register.go files that I had picked up in a rebase. Let's see if it passes e2e now.

**Comment by lavalamp** [comment]
@smarterclayton passing e2e now. PTAL!

File: pkg/apis/abac/v1beta1/register.go [hunk 3]
```diff
@@ -19,13 +19,24 @@ package v1beta1
...
+	"k8s.io/kubernetes/pkg/runtime"
 )
 
 // GroupVersion is the API group and version for abac v1beta1
 var GroupVersion = unversioned.GroupVersion{Group: api.Group, Version: "v1beta1"}
 
 func init() {
-	api.Scheme.AddKnownTypes(GroupVersion,
+	// TODO: is this really necessary?
+	addKnownTypes(api.Scheme)
```
Comment (smarterclayton) [inline pkg/apis/abac/v1beta1/register.go]: Were you getting unexpected behavior here when you removed it?

**Comment by smarterclayton** [comment]
One question, otherwise looks fine to me.

File: pkg/apis/abac/v1beta1/register.go (hunk 3 above)
Comment (lavalamp) [inline pkg/apis/abac/v1beta1/register.go]: No, didn't try--just didn't want to make any behavior changes. I will remove all of these in a future PR and deal with the fallout there.

**Comment by lavalamp** [comment]
hold off on review, this is not ready yet.

File: pkg/apis/abac/v1beta1/register.go (hunk 3 above)
Comment (lavalamp) [inline pkg/apis/abac/v1beta1/register.go]: Yes, if I remove this, abac package gives lots of test errors.

**Comment by lavalamp** [comment]
OK this is rebased and should be working again. I suggest commit-by-commit review; the second commit is big but mechanical and the others should be easy. I think I left TODOs marking most of the things that I intend to fix in follow-ups. @smarterclayton let me know if you won't get to this soon so I can find someone who can-- I don't want to go through another rebase round on this.

**Comment by nikhiljindal** [comment]
@lavalamp can you add some detail to the description. How are you making the scheme building composable? A summary of changes will be good :)

I went through some  of this pain while writting https://github.com/kubernetes/kubernetes/pull/30421, so would love to see improvements!

**Comment by lavalamp** [comment]
@nikhiljindal I added some description. Feel free to review this if you have time :)

File: pkg/api/register.go [hunk 4]
```diff
@@ -51,10 +55,27 @@ func Resource(resource string) unversioned.GroupResource {
...
-	if err := Scheme.AddIgnoredConversionType(&unversioned.TypeMeta{}, &unversioned.TypeMeta{}); err != nil {
+var (
+	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes, addDefaultingFuncs)
+	AddToScheme   = SchemeBuilder.AddToScheme
+)
+
+func init() {
+	// TODO(lavalamp): move this call to scheme builder above.
+	// Can't fix this without doing proper dependency tracking-- otherwise
+	// this will get double registered. (Consequence of not fixing this is
```
Comment (nikhiljindal) [inline pkg/api/register.go]: Why will it get double registered if moved to SchemeBuilder?

File: pkg/api/register.go [hunk 5]
```diff
@@ -51,10 +55,27 @@ func Resource(resource string) unversioned.GroupResource {
...
+	// that it only ever gets registered in the main api.Scheme, even
+	// though everyone needs these.)
```
Comment (nikhiljindal) [inline pkg/api/register.go]: Why does everyone needs these?
Seems like it works fine by adding them only in main api.Scheme?

File: pkg/apis/abac/register.go [hunk 6]
```diff
@@ -26,15 +26,27 @@ import (
 const Group = "abac.authorization.kubernetes.io"
 
 // Scheme is the default instance of runtime.Scheme to which types in the abac API group are registered.
+// TODO: remove this, abac should not have its own scheme.
```
Comment (nikhiljindal) [inline pkg/apis/abac/register.go]: Looks like we have the same issue in `federation/apis/core/register.go` and `plugin/pkg/scheduler/api/register.go`

File: pkg/apis/extensions/install/install_test.go [hunk 7]
```diff
@@ -21,6 +21,7 @@ import (
 	"testing"
 
 	"k8s.io/kubernetes/pkg/api"
+	_ "k8s.io/kubernetes/pkg/api/install"
```
Comment (nikhiljindal) [inline pkg/apis/extensions/install/install_test.go]: Add a comment saying why is this required

File: pkg/client/unversioned/clientcmd/api/latest/latest.go [hunk 8]
```diff
@@ -40,12 +40,24 @@ const OldestVersion = "v1"
...
-var Codec runtime.Codec
+var (
+	Codec  runtime.Codec
+	Scheme *runtime.Scheme
+)
 
 func init() {
-	yamlSerializer := json.NewYAMLSerializer(json.DefaultMetaFactory, api.Scheme, api.Scheme)
+	Scheme = runtime.NewScheme()
+	if err := api.AddToScheme(Scheme); err != nil {
```
Comment (nikhiljindal) [inline pkg/client/unversioned/clientcmd/api/latest/latest.go]: Should we rename this to install.go to be consistent with other code?

File: pkg/runtime/scheme_builder.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+package runtime
+
+// SchemeBuilder collects functions that add things to a scheme. It's to allow
+// code to compile without explicitly referencing generated types. You should
+// declare one in each package that will have generated deep copy or conversion
+// functions.
+type SchemeBuilder []func(*Scheme) error
+
+// AddToScheme applies all the stored functions to the scheme. A non-nil error
+// indicates that one function failed and the attempt was abandonded.
```
Comment (nikhiljindal) [inline pkg/runtime/scheme_builder.go]: abandonded -> abandoned

File: pkg/runtime/scheme_builder.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+func (sb *SchemeBuilder) AddToScheme(s *Scheme) error {
```
Comment (nikhiljindal) [inline pkg/runtime/scheme_builder.go]: AddToScheme doesnt sound right. We are running some funcs over it, which might involve adding objects or not. How about naming this ApplyToScheme or ProcessScheme?

File: plugin/pkg/scheduler/api/register.go [hunk 11]
```diff
@@ -22,19 +22,33 @@ import (
...
 var SchemeGroupVersion = unversioned.GroupVersion{Group: "", Version: runtime.APIVersionInternal}
 
+var (
+	SchemeBuilder = runtime.NewSchemeBuilder(addKnownTypes)
+	AddToScheme   = SchemeBuilder.AddToScheme
+)
+
 func init() {
-	if err := Scheme.AddIgnoredConversionType(&unversioned.TypeMeta{}, &unversioned.TypeMeta{}); err != nil {
+	if err := addKnownTypes(Scheme); err != nil {
```
Comment (nikhiljindal) [inline plugin/pkg/scheduler/api/register.go]: remove init here and move this call to install.go?

File: pkg/runtime/scheme_builder.go [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+// SchemeBuilder collects functions that add things to a scheme. It's to allow
+// code to compile without explicitly referencing generated types. You should
```
Comment (nikhiljindal) [inline pkg/runtime/scheme_builder.go]: Wasnt it the same even before?
If the generated code was there, then they had their own init() methods which were registering them.

**Comment by nikhiljindal** [comment]
I liked:
* Moving defaulting and conversion to their individual files
* Removing init methods from register.go. But I see some of them still have it. Is it possible to remove init methods from all register.go and always call AddToSchema from init() in install.go?

**Comment by nikhiljindal** [comment]
`pkg/apis/abac` and `plugin/pkg/scheduler/api/` seem to be the only ones which have an init in register.go.

File: pkg/apis/abac/register.go (hunk 6 above)
Comment (lavalamp) [inline pkg/apis/abac/register.go]: Federation might be ok, since you're doing something special. not sure about the scheduler.

File: pkg/client/unversioned/clientcmd/api/latest/latest.go (hunk 8 above)
Comment (lavalamp) [inline pkg/client/unversioned/clientcmd/api/latest/latest.go]: This one is really weird and I'm not entirely sure what's going on. Don't want to make more drastic changes at the moment.

File: pkg/runtime/scheme_builder.go (hunk 9 above)
Comment (lavalamp) [inline pkg/runtime/scheme_builder.go]: I thought that looked funny when I typed it...

File: pkg/api/register.go (hunk 5 above)
Comment (lavalamp) [inline pkg/api/register.go]: It works, but it's terrible design to have a global object that can't be reproduced hermetically for testing.

File: pkg/api/register.go (hunk 4 above)
Comment (lavalamp) [inline pkg/api/register.go]: Rewording.

File: pkg/apis/extensions/install/install_test.go (hunk 7 above)
Comment (lavalamp) [inline pkg/apis/extensions/install/install_test.go]: Removing. (this is the sort of change that adding things directly to api.Scheme in the api/register.go papers over.)

File: pkg/runtime/scheme_builder.go (hunk 10 above)
Comment (lavalamp) [inline pkg/runtime/scheme_builder.go]: The point was to replace the existing static "AddToScheme" methods in each register.go file, so I can't really change the name. I can do that in a follow-up.

File: plugin/pkg/scheduler/api/register.go (hunk 11 above)
Comment (lavalamp) [inline plugin/pkg/scheduler/api/register.go]: I will make fixes like this in a followup, if that's ok.

File: pkg/runtime/scheme_builder.go (hunk 12 above)
Comment (lavalamp) [inline pkg/runtime/scheme_builder.go]: Yes, but there was no way to construct your own runtime.Scheme (e.g., to test). Now you can get everything from a package by calling the scheme builder's AddToScheme, including the generated functions, if they are present. (adding this comment to the change description)

**Comment by lavalamp** [comment]
>  Is it possible to remove init methods from all register.go and always call AddToSchema from init() in install.go?

That's (one of) the next step(s). I'm trying not to change any actual behavior in this PR.

**Comment by lavalamp** [comment]
Changes pushed (in separate commit for easy review), PTAL

File: pkg/api/register.go (hunk 4 above)
Comment (nikhiljindal) [inline pkg/api/register.go]: Thanks. Its a lot clearer now.

**Comment by nikhiljindal** [comment]
Thanks for the fixes.
Fine to fix the remaining in follow up PRs, this PR is already huge!
Feel free to self-apply the LGTM label after squashing.
